guppy/src/graph/cargo/build.rs

```diff
@@ -219,14 +227,23 @@ impl<'a> CargoSetBuildState<'a> {
                             DependencyKind::Development,
                             target_platform,
                         ));
+            follow_target = follow_target
+                && resolver
+                    .as_mut()
+                    .map(|r| r.accept(query, link))
+                    .unwrap_or(true);
 
             // Proc macros build on the host, so for normal/dev dependencies redirect it to the host
             // instead.
             let proc_macro_redirect = follow_target && to.is_proc_macro();
```

Comment:
This should also have the filter applied to it. (Moving the check to the beginning of the callback would solve this.)